dotnet/src/Connectors/Connectors.Memory.Kusto/KustoMemoryStore.cs

```diff
@@ -0,0 +1,393 @@
+    }
+    public static MemoryRecordMetadata? DeserializeMetadata(string metadata)
+    {
+        if (string.IsNullOrEmpty(metadata))
+        {
+            return default;
+        }
+
+        return JsonConvert.DeserializeObject<MemoryRecordMetadata>(metadata);
+    }
+}
+
+internal static class KustoExtensions
+{
+    public static async ValueTask<T?> FirstOrDefaultAsync<T>(this IAsyncEnumerable<T> source, CancellationToken cancellationToken = default)
+    {
+        await foreach (var item in source.WithCancellation(cancellationToken).ConfigureAwait(false))
+        {
+            return item;
+        }
+
+        return default;
+    }
+}
+
+internal class KustoMemoryRecord
+{
+    public string Key { get; set; }
+    public MemoryRecordMetadata Metadata { get; set; }
+    public Embedding<float> Embedding { get; set; }
+    public DateTime? Timestamp { get; set; }
+
+    public KustoMemoryRecord(MemoryRecord record) : this(record.Key, record.Metadata, record.Embedding, record.Timestamp?.UtcDateTime) { }
+
+    public KustoMemoryRecord(string key, MemoryRecordMetadata metadata, Embedding<float> embedding, DateTime? timestamp = null)
+    {
+        this.Key = key;
+        this.Metadata = metadata;
+        this.Embedding = embedding;
+        this.Timestamp = timestamp;
+    }
+
+    public KustoMemoryRecord(string key, string metadata, string embedding, DateTime? timestamp = null)
+    {
+        this.Key = key;
+        this.Metadata = KustoSerializer.DeserializeMetadata(metadata);
+        this.Embedding = KustoSerializer.DeserializeEmbedding(embedding);
+        this.Timestamp = timestamp;
+    }
+
+    public MemoryRecord ToMemoryRecord()
+    {
+        return new MemoryRecord((MemoryRecordMetadata)this.Metadata, (Embedding<float>)this.Embedding, this.Key, this.Timestamp);
+    }
+
+    public string ToCsvString()
+    {
+        // Escapse string values in metadata according the Kusto csv escaping rules, whicm means replacing all \" (backslash double qoutes) with "" (two double qoutes)
+        var jsonifiedMetadata = KustoSerializer.SerializeMetadata(this.Metadata);
+        var jsonifiedEmbedding = KustoSerializer.SerializeEmbedding(this.Embedding);
+        var isoFormattedDate = this.Timestamp?.ToString("o", CultureInfo.InvariantCulture) ?? string.Empty;
+        return $"\"{this.Key}\",\"{jsonifiedMetadata}\",\"{jsonifiedEmbedding}\",{isoFormattedDate}";
+    }
+}
+
+/// <summary>
+/// An implementation of <see cref="IMemoryStore"/> backed by a Kusto database.
+/// </summary>
+/// <remarks>The embedded data is saved to the Kusto database specified in the constructor.
+/// Similarity search capability is provided through a cosine similarity function (added on first collection creation). Use Kusto's "Table" to implement "Collection".
+/// </remarks>
+public class KustoMemoryStore : IMemoryStore
+{
+    private string _database;
+    private static ClientRequestProperties GetClientRequestProperties() => new()
```

Comment:
>GetClientRequestProperties [](http://example.com/codeflow?start=43&length=26)

You really want to create a new ClientRequestId here.

Reply:
I am, just a shortened syntax for function expression.